### ISSUE TITLE
AllowNull and flow analysis conflicting design

### ISSUE THREAD

**Issue by cartermp** [dotnet/roslyn#37313]
**Version Used**:

Microsoft Visual Studio Community 2019 Int Preview
Version 16.3.0 Preview 1.0 [29116.102.d16.3]
VisualStudio.16.IntPreview/16.3.0-pre.1.0+29116.102.d16.3
Microsoft .NET Framework
Version 4.8.03752

C# Tools   3.3.0-beta1-19360-03+7d47461d2ddd32defbc51c2cf0e75a3fef16c3d7

**Steps to Reproduce**:

Consider the following example:

```csharp
using System;
using System.Diagnostics.CodeAnalysis;

namespace ConsoleApp114
{
#nullable enable

    public class Doot
    {
        private string _toot= string.Empty;

        [AllowNull]
        public string Toot
        {
            get => _toot;
            set => _toot = value ?? string.Empty;
        }
    }
    class Program
    {

        static void Main(string[] args)
        {
            var doot = new Doot();
            doot.Toot = null; // No warning, as expected
            Console.Write(doot.Toot.Length); // Unexpected warning
        }
    }
}
```

**Expected Behavior**:

No warnings.

**Actual Behavior**:

Warning on `doot.MyValue.Length`.

This is because flow analysis is mistakenly thinking we made `MyValue` a `null` value (as expected, it doesn't trace the setter, which accounts for `null`).

What I think is interesting here is that the signature of `MyValue` is saying that it allows `null` values as inputs, but does not allow `null` values for output. However, the attribute is effectively moot in this case due to flow analysis.

**Comment by safern** [comment]
cc: @stephentoub

**Comment by cartermp** [comment]
Note that if you elide the `doot.Toot = null` call here, then the last line works as expected.

**Comment by RikkiGibson** [comment]
It seems like we shouldn't track the property's flow state in this case, and just always use the property's type annotation instead. I'm just wondering what other combinations of property type annotations and flow analysis attributes should also cause us to not track a flow state.

**Comment by RikkiGibson** [comment]
Also, it seems like once we start taking flow analysis attributes on the member declaration into account when analyzing its body, we should give the setter parameter a maybe-null initial flow state in the case of `[AllowNull] string Prop { get { ... } set { ... } }`.

**Comment by jcouv** [comment]
@cartermp I think you're hitting on a more general question, unrelated to `AllowNull`: should we track state for properties that aren't trivial? Our current design says yes. We see you put a `null` in, so we assume you'll get a `null` out.
Note that you could put `NotNull` on the property to alleviate this problem.

**Comment by safern** [comment]
> Note that you could put NotNull on the property to alleviate this problem.

I'm also on the boat that when an `AllowNull` but the type is non-nullable, even if a null is passed in, it shouldn't change the state of it's nullability, because in this case we're saying, we allow null, but even if you pass null, we will not return a null value, hence the non-nullable string type in the property. 

It seems really odd to me that even though the type is saying I will not return null, the fact that I added an `AllowNull` will force me to add a `NotNull` attribute. I can't think of a scenario, where I have the combination of a non-null reference type with an `AllowNull` where I will somehow return null, so all my use cases will end up needing both attributes. Which I believe is the case for all the `AllowNull` we introduced in non-nullable non-generic types in coreclr, so that means we're missing a `NotNull` attribute in all those properties.

**Comment by Maxim-Kornilov** [comment]
@cartermp, @jcouv I faced the opposite situation when a nullable value passed by reference into a method that allows nullable value. There is no warning after invoking method `Foo`, even though the value is still might be null.

If we take a parameter by reference it's reasonable to assume that we need this in order to change the reference so that it points to another object, but keyword `ref` doesn't force us to do it like the `out` one does. I don't' have a real-world scenario for that case, but the method might have several branches of execution and only some of them might assign a new value. I think it's not safe to assume that we have a non-nullable value after invoking this method.

**Steps to Reproduce:**
```csharp
#nullable enable

using System;
using System.Diagnostics.CodeAnalysis;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            string? value = null;
            Foo(ref value);
            Console.WriteLine(value.Length); // No warning, even though the value is still might be null
        }

        static void Foo([AllowNull] ref string value)
        {
        }
    }
}
```
**Version:**
Microsoft Visual Studio Community 2019 Preview
Version 16.3.0 Preview 1.0
VisualStudio.16.Preview/16.3.0-pre.1.0+29123.89
Microsoft .NET Framework
Version 4.7.03056
C# Tools   3.3.0-beta1-19360-03+7d47461d2ddd32defbc51c2cf0e75a3fef16c3d7

**Comment by RikkiGibson** [comment]
This issue should be brought up in LDM.
